### ISSUE TITLE
Nebraska via a reverse proxy - relative URL?

### ISSUE THREAD

**Issue by MattCrum1** [kinvolk/nebraska#621]
I am working on getting updates to air-gapped Flatcar clients via Nebraska, using the host-flatcar-packages feature.

Current setup looks like this (IPs/ports are obfuscated) with a simple NGINX proxy running on our edge - we can't allow clients to directly connect to our Nebraska instance unfortunately:

![image](https://user-images.githubusercontent.com/64490589/188120656-324ceda1-297f-4b6c-93a9-c68f2eb34a91.png)

The proxy works fine for the initial update gets from the client - the target Flatcar version is returned correctly. 

When the client tries to download the package, the URL that it tries to connect to is http://**1.2.3.4:6002**/flatcar/flatcar-amd64-3227 - which it can't get to.

`
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295614  1059 omaha_request_action.cc:619] Omaha request response: <?xml version="1.0" encoding="UTF-8"?>
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: <response protocol="3.0" server="nebraska"><daystart elapsed_seconds="0"></daystart><app appid="{e96281a6-d1af-4bde-9a0a-97b76e56dc>
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295624  1059 action_processor.cc:82] ActionProcessor::ActionComplete: finished OmahaRequestAction, starting Download> ...
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295634  1059 install_plan.cc:53] InstallPlan: , new_update, url: http://1.2.3.4:6002/flatcar/flatcar-amd64-3227> ...
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295693  1059 update_attempter.cc:420] Download status: active
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295702  1059 multi_range_http_fetcher.cc:30] starting first transfer
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295711  1059 multi_range_http_fetcher.cc:58] starting transfer of range 0+?
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295724  1059 libcurl_http_fetcher.cc:47] Starting/Resuming transfer
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295804  1059 libcurl_http_fetcher.cc:151] Setting up curl options for HTTP
Sep 02 10:05:21 azure-test-hardened update_engine[1059]: I0902 10:05:21.295951  1059 libcurl_http_fetcher.cc:434] Setting up timeout source: 1 seconds.
Sep 02 10:05:32 azure-test-hardened update_engine[1059]: I0902 10:05:32.625845  1059 prefs.cc:51] certificate-report-to-send-update not present in /var/lib/update_engine/prefs
Sep 02 10:05:32 azure-test-hardened update_engine[1059]: I0902 10:05:32.625921  1059 prefs.cc:51] certificate-report-to-send-download not present in /var/lib/update_engine/prefs
Sep 02 10:05:51 azure-test-hardened update_engine[1059]: E0902 10:05:51.625787  1059 libcurl_http_fetcher.cc:250] Unable to get http response code: Connection timeout after 30330 ms
Sep 02 10:05:51 azure-test-hardened update_engine[1059]: I0902 10:05:51.625888  1059 libcurl_http_fetcher.cc:281] Transfer resulted in an error (0), 0 bytes downloaded
`

I tried setting the -nebraska-url parameter to the Reverse Proxy URL http://10.0.0.1:6999 but this broke everything on the Nebraska front-end (I assume this URL is used for auth as well as serving updates).

Question is, is there a way to set the "url codebase" string returned by Nebraska to be either a relative URL (../../flatcar), to use the HTTP host header from the client, or to somehow hardcode it without using the -nebraska-url parameter and breaking things?

Please let me know if you have any ideas.

**Comment by pothos** [comment]
> Please let me know if you have any ideas.

I think a hostname would help to have the reverse proxy be serving the same hostname or being able to rewrite the hostname in the reverse proxy.

**Comment by MattCrum1** [comment]
Yeah that would partially sort it; having an internally resolvable hostname, with an entry in /etc/hosts on the clients to override the IP... but that does not work for ports. Everything would need to be switched to run on the same port, or I get the reverse proxy to translate it somehow

**Comment by pothos** [comment]
NGINX reverse proxies are powerful but a bit hard to configure ;) In case it helps, you can also use a regular HTTP proxy: https://www.flatcar.org/docs/latest/setup/releases/update-strategies/#updating-behind-a-proxy

**Comment by MattCrum1** [comment]
Been a bit quiet on this for a while but I got things working via a reverse proxy. Proxied updates work fine; but Nebraska is showing all nodes behind the proxy with the same IP address. 

Makes sense (that's where the connection is initiated), but there's no way to identify which node is which - is there any way to pass the hostname, or the remote node's IP via the update protocol rather than basing it on the HTTP connection?

Let me know if you need any more details.

**Comment by pothos** [comment]
Each node identifies itself with the machine id from `/etc/machine-id` which must be unique. You can add human understandable context in the `MACHINE_ALIAS` entry: https://www.flatcar.org/docs/latest/setup/releases/update-conf/
Put `MACHINE_ALIAS=something` as a new line in `/etc/flatcar/update.conf`.

> 
>     Optional human-friendly name for the machine in addition to the machine ID from /etc/machine-id, to be displayed in the update server UI, should be unique but this is not enforced, use quotes if it contains whitespace
>     Set this dynamically by running, e.g., sudo sed -i "/MACHINE_ALIAS=.*/d" /etc/flatcar/update.conf ; echo "MACHINE_ALIAS=$(hostname)" | sudo tee -a /etc/flatcar/update.conf for the output of the hostname command (as with the other variables, restarting update-engine.service is not needed)
>

**Comment by MattCrum1** [comment]
Ah awesome - I'll give that a go, thanks @pothos
